kube-runtime/src/controller.rs

```diff
@@ -92,6 +99,75 @@ where
     })
 }
 
+/// A policy for when to retry reconciliation, after an error has occurred.
+pub trait ErrorPolicy {
+    type Err;
+    type K: RuntimeResource;
+    type Ctx;
+
+    /// Notifies that the state for an object should be removed, for example if it is reconciled successfully.
+    fn reset_object(&mut self, obj_ref: &ObjectRef<Self::K>, ctx: Context<Self::Ctx>);
+    /// Queries for when to next retry after an error.
+    fn on_error(
+        &mut self,
+        obj_ref: ObjectRef<Self::K>,
+        error: &Self::Err,
+        ctx: Context<Self::Ctx>,
+    ) -> ReconcilerAction;
+}
+
+/// Retries errors based on a `Backoff` policy.
+///
+/// A separate backoff tracker is used for each object, and it is
+/// reset whenever the object is reconciled successfully.
+#[derive(Debug)]
+pub struct BackoffErrorPolicy<MkBackoff, B, K: RuntimeResource, Err, Ctx> {
+    make_backoff: MkBackoff,
+    backoffs: HashMap<ObjectRef<K>, B>,
```

Comment:
Ahhh. One back off per object. Interesting. I thought we would have one global one. This makes sense though. Probably smarter.

Reply:
I /think/ we should encourage backoff-per-object by default. There are cases where global backoffs make sense (gating access to some central resource for example), but you probably want to gate those by specific error cases..